File: src/org/simbrain/network/synapse_update_rules/spikeresponders/ConvolvedJumpAndDecay.java
package org.simbrain.network.synapse_update_rules.spikeresponders;

import org.simbrain.network.core.Synapse;
import org.simbrain.util.UserParameter;

/**
 * @author Zoë Tosi
 */
public class ConvolvedJumpAndDecay extends SpikeResponder {

    /**
     * Jump height value.
     */
    @UserParameter(label = "Jump Height",
            description = "This value is multiplied by the strength to determine the total instantaneous rise in a"
                    + " post-synaptic response to an action potential or spike.",
            defaultValue = "1", order = 1)
    private double jumpHeight = 1;

    /**
     * Base line value.
     */
    @UserParameter(label = "Base-Line",
            description = "The post-synaptic response value when no spike have occurred. Alternatively, the "
                    + "post synaptic response to which decays to over time.",
            defaultValue = "0.0001", order = 1)
    private double baseLine = .0001;

    /**
     * Rate at which synapse will decay (ms).
     */
    @UserParameter(label = "Time Constant",
            description = "The time constant of decay and recovery (ms).",
            defaultValue = "3", order = 1)
    private double timeConstant = 3;

    public ConvolvedJumpAndDecay(){

    }

    public ConvolvedJumpAndDecay(double tau) {
        this.timeConstant = tau;
    }

    @Override
    public ConvolvedJumpAndDecay deepCopy() {
        ConvolvedJumpAndDecay jad = new ConvolvedJumpAndDecay();
        jad.setBaseLine(this.getBaseLine());
        jad.setJumpHeight(this.getJumpHeight());
        jad.setTimeConstant(this.getTimeConstant());
        return jad;
    }

    @Override
    public void update(final Synapse s) {
        value = s.getPsr();
        if (s.getSource().isSpike()) {
            value += jumpHeight * s.getStrength();
        } else {
            double timeStep = s.getParentNetwork().getTimeStep();
            value += timeStep * (baseLine - value) / timeConstant;
        }
        s.setPsr(value);
    }

    public void update(final Synapse s, double jump) {
        value = s.getPsr();
        if (s.getSource().isSpike()) {
            value += jump;
        } else {
            double timeStep = s.getParentNetwork().getTimeStep();
            value += timeStep * (baseLine - value) / timeConstant;
        }
        s.setPsr(value);
    }

    @Override
    public String getDescription() {
        return "Convolved Jump and Decay";
    }

    @Override
    public String getName() {
        return "Convolved Jump and Decay";
    }

    public double getBaseLine() {
        return baseLine;
    }

    public void setBaseLine(final double baseLine) {
        this.baseLine = baseLine;
    }

    public double getJumpHeight() {
        return jumpHeight;
    }

    public void setJumpHeight(final double jumpHeight) {
        this.jumpHeight = jumpHeight;
    }

    /**
     * @return the time constant of the exponential decay of the post synaptic
     * response
     */
    public double getTimeConstant() {
        return timeConstant;
    }

    /**
     * @param decayTimeConstant the new time constant of the exponential decay
     *                          of the post synaptic response
     */
    public void setTimeConstant(double decayTimeConstant) {
        this.timeConstant = decayTimeConstant;
    }

}

File: src/main/kotlin/org/simbrain/network/spikeresponders/ConvolvedJumpAndDecay.kt
package org.simbrain.network.spikeresponders

import org.simbrain.network.core.*
import org.simbrain.network.util.MatrixDataHolder
import org.simbrain.network.util.ScalarDataHolder
import org.simbrain.network.util.SpikingMatrixData
import org.simbrain.util.UserParameter

/**
 * Each time a spike occurs jump by a given value (the synaptic strength) and then decay.
 *
 * @author Zoë Tosi
 */
class ConvolvedJumpAndDecay() : SpikeResponder() {

    /**
     * Base line value.
     */
    @UserParameter(
        label = "Base-Line",
        description = "Decays to this value when no spikes occur.",
        increment = .1,
        order = 1
    )
    var baseLine = 0.0

    @UserParameter(
        label = "Time Constant",
        description = "Time constant of decay (ms). Roughly the time it takes to decay to\n" +
                "near-baseline. Larger time constants produce slower decay.",
        increment = .1,
        order = 2
    )
    var timeConstant = 3.0

    override fun copy(): ConvolvedJumpAndDecay {
        val jad = ConvolvedJumpAndDecay()
        jad.baseLine = baseLine
        return jad
    }

    context(Network)
    override fun apply(connector: Connector, responderData: MatrixDataHolder) {
        val wm = connector.let { if (it is WeightMatrix) it else return }
        val na = connector.source.let { if (it is NeuronArray) it else return }
        val spikeData = na.dataHolder.let { if (it is SpikingMatrixData) it else return }
        if (na.updateRule.isSpikingRule) {
            for (i in 0 until wm.weightMatrix.nrow()) {
                for (j in 0 until wm.weightMatrix.ncol()) {
                    val psr = convolvedJumpAndDecay(
                        spikeData.spikes[j],
                        wm.psrMatrix[i, j],
                        wm.weightMatrix[i, j],
                        timeStep
                    )
                    wm.psrMatrix.set(i, j, psr)
                }
            }
        }
    }


    context(Network)
    override fun apply(synapse: Synapse, responderData: ScalarDataHolder) {
        synapse.psr = convolvedJumpAndDecay(synapse.source.isSpike, synapse.psr, synapse.strength, timeStep)
    }

    fun convolvedJumpAndDecay(
        spiked: Boolean,
        psr: Double,
        jumpHeight: Double,
        timeStep: Double): Double {
        return if (spiked) {
            psr + jumpHeight
        } else {
            psr + timeStep * (baseLine - psr) / timeConstant
        }
    }

    override val description = "Convolved Jump and Decay"

    override val name = "Convolved Jump and Decay"
}